File: moderate/pile_of_bricks.hs
import System.Environment (getArgs)
import Data.List.Split (splitOn)
import Data.List (intercalate, sort)

clean   :: String -> String
clean xs = [x | x <- xs, notElem x "()[]"]

pileofbric            :: [Int] -> [[String]] -> [String]
pileofbric _  []       = []
pileofbric hs (xs:xss) | head b <= head hs && b!!1 <= hs!!1 = head xs : pileofbric hs xss -- || diagonal b hs
                       | otherwise                          = pileofbric hs xss
                       where t = map read (splitOn "," (xs!!1))
                             u = map read (splitOn "," (xs!!2))
                             b = sort [abs (head t - head u), abs (t!!1 - u!!1), abs (t!!2 - u!!2)]

pileofbrick             :: [String] -> [String] -> [String]
pileofbrick [xs, ys] zss = pileofbric h (map words zss)
                         where s = map read (splitOn "," xs)
                               t = map read (splitOn "," ys)
                               h = sort [abs (head s - head t), abs (s!!1 - t!!1)]

pileofbricks         :: [String] -> String
pileofbricks [xs, ys] | null p    = "-"
                      | otherwise = intercalate "," r
                      where p = pileofbrick (words xs) (splitOn ";" ys)
                            r = map show (sort (map (read :: String -> Int) p))

main :: IO ()
main = do
    [inpFile] <- getArgs
    input <- readFile inpFile
    putStr . unlines . map (pileofbricks . splitOn "|") . lines $ clean input

File: moderate/pile_of_bricks.lua
for line in io.lines(arg[1]) do
  h, b, r = {}, {}, {}
  s1, s2, s3, s4 = line:match("%[([-]?%d+),([-]?%d+)%] %[([-]?%d+),([-]?%d+)%]")
  h[1], h[2] = math.abs(tonumber(s1)-tonumber(s3)), math.abs(tonumber(s2)-tonumber(s4))
  table.sort(h)
  for i in line:gmatch("%d+ %[[-]?%d+,[-]?%d+,[-]?%d+%] %[[-]?%d+,[-]?%d+,[-]?%d+%]") do
    ix, s1, s2, s3, s4, s5, s6 = i:match("(%d+) %[([-]?%d+),([-]?%d+),([-]?%d+)%] %[([-]?%d+),([-]?%d+),([-]?%d+)%]")
    b[1], b[2], b[3] = math.abs(tonumber(s1)-tonumber(s4)), math.abs(tonumber(s2)-tonumber(s5)), math.abs(tonumber(s3)-tonumber(s6))
    table.sort(b)
    if b[1] <= h[1] and b[2] <= h[2] then -- or diagonal(b, h)
      r[#r + 1] = tonumber(ix)
    end
  end
  if #r > 0 then
    table.sort(r)
    print(table.concat(r, ","))
  else
    print("-")
  end
end
